airbyte-integrations/bases/base-java/src/main/java/io/airbyte/integrations/base/IntegrationRunner.java

```diff
@@ -111,77 +117,102 @@ private void runInternal(final IntegrationConfig parsed) throws Exception {
-                            .withMessage(displayMessage)));
-        return;
-      }
-      throw e;
+      handleRunInternalException(parsed, e);
     }
 
     LOGGER.info("Completed integration: {}", integration.getClass().getName());
   }
 
+  private AirbyteMessage runCheck(final IntegrationConfig parsed) throws Exception {
+    try {
+      final JsonNode config = getValidatedConfig(parsed, CHECK.toString());
+      return new AirbyteMessage().withType(Type.CONNECTION_STATUS).withConnectionStatus(integration.check(config));
+    } catch (final ConfigValidationException e) {
+      // if validation fails don't throw an exception, return a failed connection check message
+      return failedConnectionStatusMessage(e.getMessage());
+    }
+  }
+
+  private static boolean isCheckWithCatalog(IntegrationConfig parsed) {
+    return CHECK.equals(parsed.getCommand()) && parsed.getCatalogPath() != null;
+  }
+
+  private AirbyteMessage runCheckWithCatalog(final IntegrationConfig parsed) throws Exception {
+    LOGGER.info("Using Check With Catalog Method");
+    // TODO: Actually implement something different
+    return runCheck(parsed);
+  }
+
+  private AirbyteMessage runDiscover(final IntegrationConfig parsed) throws Exception {
+    final JsonNode config = getValidatedConfig(parsed, DISCOVER.toString());
+    return new AirbyteMessage().withType(Type.CATALOG).withCatalog(source.discover(config));
+  }
+
+  private void runRead(final IntegrationConfig parsed) throws Exception {
+    final JsonNode config = getValidatedConfig(parsed, READ.toString());
+    final ConfiguredAirbyteCatalog catalog = parseConfig(parsed.getCatalogPath(), ConfiguredAirbyteCatalog.class);
+    final Optional<JsonNode> stateOptional = parsed.getStatePath().map(IntegrationRunner::parseConfig);
+    try (final AutoCloseableIterator<AirbyteMessage> messageIterator = source.read(config, catalog, stateOptional.orElse(null))) {
+      produceMessages(messageIterator);
+    }
+  }
+
+  private void runWrite(final IntegrationConfig parsed) throws Exception {
+    final JsonNode config = getValidatedConfig(parsed, WRITE.toString());
+    final ConfiguredAirbyteCatalog catalog = parseConfig(parsed.getCatalogPath(), ConfiguredAirbyteCatalog.class);
+    try (final AirbyteMessageConsumer consumer = destination.getConsumer(config, catalog, outputRecordCollector)) {
+      runConsumer(consumer);
+    }
+  }
+
+  private JsonNode getValidatedConfig(final IntegrationConfig parsed, final String operationType) throws Exception {
+    final JsonNode config = parseConfig(parsed.getConfigPath());
+    validateConfig(integration.spec().getConnectionSpecification(), config, operationType);
+    return config;
+  }
+
+  private void handleRunInternalException(final IntegrationConfig parsed, final Exception e) throws Exception {
+    // Many of the exceptions thrown are nested inside layers of RuntimeExceptions. An attempt is made
+    // to
```

Comment:
Minor nit: since we're already refactoring this, can we potentially make this comment not have a single word comment line? 😆

Reply:
good call